Add project

ajeddeloh opened the pull request:
Huge apologies to anyone who was using this unbeknownst to me. Wanted to have this actually be reviewed first.

I'd prefer to not rewrite a bunch of history for this, so unless there's a good reason I'd prefer to just add commits to this as it's reviewed.

cc @jlebon since he looked at this before

Commit:
*: begin project

Commit:
json: add json decoder

Commit:
path: use correct package

Commit:
path: add line and column functions

Commit:
readme: add readme

Commit:
yaml: add yaml support

Known bug: offset/line is not correct. Will be fixed in later commit
reworking Marker

Commit:
tree: mv path to tree, clean up interface

Move path.go to a more descriptive name. Clean up it's interface in
preperation for validation code using it.

Commit:
path/: add path package

Add a package for describing a json or yaml path.

Commit:
report: add report package

Add package for describing a set of errors or warnings from a config.
This roughly mirrors the Ignition config/validate/report package.

Commit:
validate: add package

Add package for validating go structures. This roughly mirrors the
Ignition config/validate package but does not handle unused keys or
line/col reporting. Those will be handled seperately.

Commit:
tree.go: split mark into pos, add way to get lines

Split Marker into a set of Pos's. Add a way to fill in line/column from
a source blob based on indices.

Commit:
report: add AddOnError function

Commit:
tree: move to own package

Additionally add some more helpers in report.

Commit:
validate: fix nil pointer dereference

Commit:
report: fix nil pointer dereference

Commit:
report: add correlate function

Add function to correlate context in a report with line/column
information.

Commit:
report: relax correlate() to not error

Make report.Correlate() never return an error. If an report entry's
context does not exist in the tree, keep trying with less context until
you reach the root.

Commit:
tree: change Get() to take a list, not varargs

Commit:
tree: don't add negative offsets, fix index oob

Skip adding Pos's to Markers when the index is negative. Fix index oob
in FixLineColumn.

Commit:
test: add test script, fix gofmt

Commit:
validate: add tests

Add tests for running the validation code. Additionally fix bugs with
the the validation code the tests uncovered.

Commit:
readme: update and rename

Commit:
*: apply ASL 2.0 license

Commit:
tree: add tests for FixLineColumn

Additionally refactor it a bit to be easier to test.

Commit:
tree: test duplicate positions, fix code for test

Test to ensure having duplicate Pos's (by index) doesn't cause problems.
Fix the code to not have problems with it.

Commit:
path: include tag in path

The ContextPath should contain the metadata about what kind of source it
came from. This allows third party code to make use of it in Validate()
functions if they desire and cleans up the validation functions a bit.

Commit:
validate: allow custom validation

All specifying custom validation functions to run on the reflect.Values
that we walk that aren't based on the validator interface. This is
useful for implementing things like duplicate checking or checking for
unused keys.

jlebon committed:
README.md: fix typos and tweak example

Just random things I noticed while perusing the codebase.

Commit:
report,tree: improve comments

Commit:
path: add New() function to help with verbosity

Commit:
validate: add test for GetFields

This is converted from the old test in Ignition.

Commit:
report: make AddOn() public

Commit:
test,go.mod: allow go 1.11, clean up test

Commit:
json: add tests for fromJsonNode

Additionally fix go.sum by running go mod tidy

jlebon commented on report/report.go:
```diff
@@ -21,15 +21,18 @@ import (
 	"github.com/ajeddeloh/vcontext/tree"
 )
 
+// EntryKind represents an Entry's severity
```
Minor: docstrings in this patchset have inconsistent final dot.

ajeddeloh replied:
Fixed

jlebon reviewed (commented):
Only checked commits after when I had last looked at this repo. Definitely appreciate the added documentation! :+1: 

Would be good to have someone else glance over the rest of the code too though. At least at an architectural level.

Commit:
*: fix inconsistent comment punctuation

arithx commented on report/report.go:
```diff
@@ -0,0 +1,153 @@
+		at = fmt.Sprintf("at %s", e.Context.String())
+	}
+
+	return fmt.Sprintf("%s %s: %s", e.Kind.String(), at, e.Message)
```
Might be worthwhile to either have a default case that populates `at` with some indicator about not having positional context or to move the space between `e.Kind.String()` & `at` from this format statement to the beginning of the `at` string (in cases that it has a value).

Right now if it somehow fell through the `switch` without hitting any of the cases the output message would be `<kind> : <msg>` with an extra space before the `:`.

arithx commented on report/report.go:
```diff
@@ -0,0 +1,153 @@
+	case Info:
+		return "info"
+	default:
+		return ""
```
Should this default to something like `unknown`? Seems kind of odd that an entry could be in an error state and we just wouldn't report anything about the `Kind`.

ajeddeloh replied:
I don't like `unknown` since it's not clear to a user what exactly is unknown and depending on what the error message is it could be misleading. Panicing seemed like the wrong answer and returning `(string, error)` would be a pain. Maybe `unknown severity`?

arithx replied:
`unknown severity` wfm.

arithx commented on validate/validate.go:
```diff
@@ -0,0 +1,149 @@
+}
+
+// GetFields takes a value of a struct and flattens all embedded structs in it.
+// If any fields are interfaces, it "dereferences" them interface to the underlying type.
```
`s,them,the,g`

arithx reviewed (commented):
Is it intentional to skip the `vendor/` directory and `-mod=vendor` on the `go test`?

ajeddeloh commented:
regarding the `vendor/` directory, I think I'm fine with leaving it off since there's only two deps and I maintain one of them.

Commit:
report: drop extra space with no context/marker

Generate entry string like "error: message" not "error : message" when
we don't have a logic (path-like) or line/column location of the error.

Commit:
report: report "unknown severity" instead of ""

Be explicit that the severity is unknown instead of omiting the
severity.report: report "unknown severity" instead of ""

Commit:
validate: fix mistake in comment

arithx reviewed (approved):
LGTM
